### ISSUE TITLE
Memory Leak in crossbeam-queue ArrayQueue? (Latest git only, ver0.2.3 is not effected)

### ISSUE THREAD

**Issue by YoshikiTakashima** [crossbeam-rs/crossbeam#539]
Hi.

I'm Yoshi, a PhD student at CMU. We are currently evaluating a project that aims to automatically synthesize test cases for Rust libraries.

We ran the synthesized test cases with Miri, and it reports what appears to be a memory leak. The minimum example is:
```
{
let x_1 = ArrayQueue::<i32>::new(1);
}//destruct x_1
```
```
error: Undefined Behavior: incorrect layout on deallocation: allocation has size 64 and alignment 8, but gave size 16 and alignment 8
```
It looks like the drop destructor is not erasing enough memory.

You can also catch this with by running `cargo miri test` inside crossbeam-queue. One of your manually written test cases (`capacity`) will induce the same behavior.

The first commit to have this issue is [2180b820c846e5e6f2f11ccd1238198a5861dc08](https://github.com/crossbeam-rs/crossbeam/commit/2180b820c846e5e6f2f11ccd1238198a5861dc08). You can confirm this by running `cargo miri test` on 1 commit before. It will not terminate because Miri is extremely slow, but you can see that it gets past `capacity` test case.

As for the root cause, I suspect that the drop function needs to be modified to match the changes to `new`, but I am yet to get a PR up.

In case this is intended behavior, or you would prefer if I focused on other parts of the code, please let me know.

Thanks
~Yoshi

**Comment by cynecx** [comment]
Yes, I just realized that the changes in https://github.com/crossbeam-rs/crossbeam/pull/500 is not correct in terms of equivalence compared to the original code behavior.

The PR (#500) replaces:

```rust
let mut v = Vec::<Slot<T>>::with_capacity(cap);
```

with

```rust
let mut v: Vec<Slot<T>> = (0..cap)
    .map(|i| {
        // Set the stamp to `{ lap: 0, mark: 0, index: i }`.
        Slot {
            stamp: AtomicUsize::new(i),
            msg: UnsafeCell::new(MaybeUninit::uninit()),
        }
    })
    .collect();
```

The latter does **not** exactly allocate `cap` slot entries as `with_capacity` would do, instead with `cap=1` the current `Vec` + `collect` implementation would allocate 4 entries. When the `drop` implementation [1] would run, it would try to reconstruct the `Vec` with the `cap` as capacity although it is not the correct capacity, hence causing the memory leak.

[1]:

https://github.com/crossbeam-rs/crossbeam/blob/5a68889cbc490f71fce5e615572e586cf15668a6/crossbeam-queue/src/array_queue.rs#L426-L429

**Comment by caelunshun** [comment]
This is fixed with #533. Note that the same issue was also present before #500 was merged, since Vec::with_capacity doesn't guarantee that exact capacity.

**Comment by cynecx** [comment]
@caelunshun Ah I didn't see that PR. 

> since Vec::with_capacity doesn't guarantee that exact capacity.

The std docs says: "The vector will be able to hold exactly capacity elements without reallocating. If capacity is 0, the vector will not allocate.". Ährm, doesn't that kinda say that the capacity of the Vec will be exactly `cap`?

**Comment by caelunshun** [comment]
Ah, looks like I was mistaken there. You're right, then; this issue was introduced in #500.

**Comment by cynecx** [comment]
@caelunshun Mhh. Maybe a simpler fix for #500 would have been something like:

```rust
// Instead of
let mut v = Vec::<Slot<T>>::with_capacity(cap);
// do this
let mut v = Vec::<MaybeUninit<Slot<T>>>::with_capacity(cap);
```

So that the actual initialization loop can be preserved.

**Comment by YoshikiTakashima** [comment]
Hi @cynecx, @caelunshun.

Thank you for the quick response.

I'll merge  [#533](https://github.com/crossbeam-rs/crossbeam/pull/533) into my fork and see if I can find more interesting issues.

Thank you for the help!
